---
title: "EpiGIFT_Seurat_Extract_FindMarkers_List"
author: "Laëtitia Racine"
date: "2023-03-20"
subtitle: "Last modification : `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    code_folding: "hide"
    toc: true
    toc_float: false
    theme: journal
---

<style>
body {text-align: justify}
div.beige {background-color:#FFE0B2; border-radius: 5px; padding: 8px;}
</style>

```{r, Setup, include=FALSE}

knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(warning = FALSE)
knitr::opts_chunk$set(message = FALSE)

```

```{r, Dependencies}

library(dplyr)
library(Seurat)
library(stringr)
library(tidyverse)
library(kableExtra)

```

```{r, Working directories and external script}

start_time = Sys.time()
directory = str_extract(string = getwd(), pattern = "[:graph:]+(?=bin)")
directory_data = paste0(directory, "data/")
current_date = format(Sys.time(), "%Y%m%d")
dir.create(path = paste0(directory, "exp/scRNAseq_EpiGIFT_Seurat_FindMarkers_ListGenes/"))
dir.create(path = paste0(directory, "exp/scRNAseq_EpiGIFT_Seurat_FindMarkers_ListGenes/", current_date))
directory_output = paste0(directory, "exp/scRNAseq_EpiGIFT_Seurat_FindMarkers_ListGenes/", current_date, "/")

```

****

The data we use are part of a big dataset of 10XscRNAseq together with CITEseq from 10 conditions (cf. Git_Differentiation_Metabolism, output of the scRNAseq_CITEseq_R_NormalizeData.R script before the SCT normalization). Here, we only use the cells from two conditions at the time. The whole dataset was already analyzed using the standard Seurat method. Two filters were applied on cells (keep only cells with 1000 to 6500 genes detected and keep only cells with less than 25% of mRNA and more than 5% of rRNA) and two filters were applied on genes (remove mitochondrial genes and remove genes detected in less than 3 cells). 



<br><br><br>



# Script's objective 

<br>

Extracting list of differentially expressed genes between two conditions using the FindMarkers() function from Seurat package.   

```{r, Input loading}

list_seurat_obj = readRDS(paste0(directory_data, "list_seurat_obj_epi.rds"))

```

Nb : the first part of this script is a copy of what we can find in *scRNAseq_EpiGIFT_Seurat_InvestigationGenesSeurat.Rmd*. We decided to clean the code in here to keep only the chosen parameters and function.  


<br><br><br>



# FindMarkers function explanation

<br>

Documentation :    
https://satijalab.org/seurat/reference/findmarkers      
https://github.com/satijalab/seurat/blob/96d07d80bc4b6513b93e9c10d8a9d57ae7016f9f/R/differential_expression.R#L176-L178        
https://github.com/satijalab/seurat/issues/467      
https://satijalab.org/seurat/articles/v4_changes.html      
       
The function has some interesting parameters that we need to look at :    
- logfc.threshold = 0.25 : limit testing to genes which show, on average, at least X-fold difference (log2-scale !!! change in Seurat v4) between the two groups of cells    
- max.cells.per.ident = -Inf : no downsampling on the cells      
- features = NULL : use all genes to test => *list of variable features has no consequence here*   
- min.pct = 0.1 : only test genes that are detected in a minimum fraction of min.pct cells in either of the two populations    
- test.use = wilcox : test to use for the differential analysis. Lots of options are available here : wilcox (rank sum test), bimod (likelihood-ratio test), roc, t-test, negbinom, poisson, LR (logistic regression), MAST (hurdle model) and DESeq2 (negative binomial distribution). See documentation for more informations.    
  
Find markers function's code :    
1) Matrix with cells from ident.1 is extracted        
2) Matrix with cells from ident.2 is extracted      
3) expm1 (exponential minus 1) is calculated for each cell    
4) mean of the expm1 is calculated for each gene      
5) the pseudocount number (default = 1) is added to the mean      
6) the log2 of the value is calculated    
=> so at this step, we have one value per gene per ident  
7) the difference between the value of ident.1 and ident.2 is calculated, this is the log2FC 
8) name of the genes with abs(log2FC) > logFC.threshold are saved in a list genes.use    
=> all those steps allow to filter the initial matrix because all the genes are not kept in the statistical test    
9) the initial matrix is filtered to keep only genes from genes.use    
10) a statistical test (default = wilcoxon) is performed    
11) the output of the test is a table with pvalue    
12) bonferroni correction is applied to calcul the p value adjusted (bonferroni : multiply the pval by the number of comparison (ie. nb genes in the seurat object = 22084))
  
We obtain as an *output* a data.frame with a ranked list of putative markers as rows, and associated statistics as columns (p-values, ROC score, etc., depending on the test used (test.use)). The following columns are always present:  
    • avg_log2FC: log2 fold-change of the average expression between the two groups. Positive values indicate that the gene is more highly expressed in the first group  
    • pct.1: The percentage of cells where the gene is detected in the first group  
    • pct.2: The percentage of cells where the gene is detected in the second group  
    • p_val_adj: Adjusted p-value, based on bonferroni correction using all genes in the dataset  

<br>

```{r, Function definition}

extract_list_FM = function(list_seurat, drug_choice){
  
  list_seurat = list_seurat[drug_choice]
  
  # Apply SCT normalization on individual objects
  list_obj = lapply(list_seurat, function(x) { 
    SCTransform(object = x, 
                assay = "RNA", 
                new.assay.name = "SCT",
                do.center = TRUE, 
                do.scale = FALSE,
                variable.features.n = NULL,
                variable.features.rv.th = 1.3,
                residual.features = NULL,
                vars.to.regress = NULL,
                min_cells = 1,
                ncells = NULL,
                verbose = FALSE,
                # vst additional parameter
                method = "glmGamPoi",
                residual_type = "deviance",
                n_genes = NULL)
  })

  # Merge conditions
  merge_obj = merge(x = list_obj[[drug_choice[1]]], 
                    y = list_obj[[drug_choice[2]]],
                    add.cell.ids = drug_choice,
                    merge.data = TRUE)
  
  # Variable features
  VariableFeatures(merge_obj) = rownames(merge_obj[["SCT"]]@scale.data)
  
  # Linear reduction and clustering
  merge_obj = RunPCA(object = merge_obj, 
                     assay = "SCT", 
                     npcs = 100,
                     reduction.name = "pca", 
                     seed.use = 42, 
                     verbose = FALSE)
  merge_obj = RunUMAP(object = merge_obj, 
                      assay = "SCT", 
                      dims = c(1:40), 
                      set.seed = 42,
                      reduction =  "pca", 
                      reduction.name = "umap", 
                      verbose = FALSE)
  merge_obj = FindNeighbors(object = merge_obj, assay = "SCT", dims = c(1:40))
  merge_obj = FindClusters(object = merge_obj, resolution = 0.6, algorithm = 3)
  
  # Prepare the object for FindMarkers
  Idents(merge_obj) <- "orig.ident"
  merge_obj <- PrepSCTFindMarkers(merge_obj)

  # FindMarkers function application
  genes = FindMarkers(object = merge_obj, 
                      assay = "SCT", 
                      ident.1 = drug_choice[1], 
                      ident.2 = drug_choice[2], 
                      logfc.threshold = 0.15,
                      slot = "data",
                      group.by = NULL,
                      subset.ident = NULL,
                      features = NULL,
                      test.use = "wilcox",
                      pseudocount.use = 1,
                      verbose = FALSE) %>% 
      tibble::rownames_to_column("gene") %>%
      dplyr::filter(p_val_adj <= 0.05)
    
    return(genes)
}

```

<br>

<div class = "beige">

Chosen parameters :   
- logfc.threshold = 0.15    
- statistical test = wilcoxon rank sum test  
- pvalue adjusted (bonferonni) threshold = 0.05  
https://data.library.virginia.edu/the-wilcoxon-rank-sum-test/  
https://www.statisticssolutions.com/bonferroni-correction/  

</div>



<br><br><br>



# Extract list of genes for each conditions combinations (CTRL vs DRUG)

<br>

**Choose conditions**
```{r}

drugs_1 = c("CTRL", "DON")
drugs_2 = c("CTRL", "2DG")
drugs_3 = c("CTRL", "AOA")

```

<br>


**Call the function**
```{r}

genes_tab1 = extract_list_FM(list_seurat_obj, drugs_1)
genes_tab2 = extract_list_FM(list_seurat_obj, drugs_2)
genes_tab3 = extract_list_FM(list_seurat_obj, drugs_3)

```

<br>


**Extract list of genes**
```{r}

genes_list1 = genes_tab1$gene
genes_list2 = genes_tab2$gene
genes_list3 = genes_tab3$gene

cat("Number of genes for", drugs_1, length(genes_list1), "genes \n",
    "Number of genes for", drugs_2, length(genes_list2), "genes \n",
    "Number of genes for", drugs_3, length(genes_list3), "genes \n")

```


<br>

**Save outputs**  
```{r}

# Genes names as list 
write.csv2(x = genes_list1, row.names = FALSE,
           file = paste0(directory_output, paste(drugs_1, collapse = "-"), "_list_", 
                         length(genes_list1), "_genes_FindMarkers_th0.15_wilcox_pval0.05.csv"))
write.csv2(x = genes_list2, row.names = FALSE, 
           file = paste0(directory_output, paste(drugs_2, collapse = "-"), "_list_", 
                         length(genes_list2), "_genes_FindMarkers_th0.15_wilcox_pval0.05.csv"))
write.csv2(x = genes_list3, row.names = FALSE,
           file = paste0(directory_output, paste(drugs_3, collapse = "-"), "_list_", 
                         length(genes_list3), "_genes_FindMarkers_th0.15_wilcox_pval0.05.csv"))

```

```{r}

# Genes names and FindMarkers information as a table
saveRDS(genes_tab1, file = paste0(directory_output, paste(drugs_1, collapse = "-"),
                                  "_list_", length(genes_list1), 
                                  "_genes_FindMarkers_th0.15_wilcox_pval0.05.rds"))
saveRDS(genes_tab2, file = paste0(directory_output, paste(drugs_2, collapse = "-"), 
                                  "_list_", length(genes_list2), 
                                  "_genes_FindMarkers_th0.15_wilcox_pval0.05.rds"))
saveRDS(genes_tab3, file = paste0(directory_output, paste(drugs_3, collapse = "-"), 
                                  "_list_", length(genes_list3), 
                                  "_genes_FindMarkers_th0.15_wilcox_pval0.05.rds"))

```



<br><br><br>



****

```{r, Rsession}

end_time = Sys.time()
cat("Total execution time : ", as.numeric (end_time - start_time, units = "mins"), "minutes")

# Show package version
sessionInfo()

```

```{r, results='hide'}

# Clean working space and memory 
rm(list = ls())
gc()

```